filecoin-proofs/src/pieces.rs

```diff
@@ -110,6 +110,64 @@ pub fn get_aligned_source(
     )
 }
 
+/// Insert piece paddings, such that the final result includes _all_ pieces that need to be hashed
+/// to generate a full commitment.
+pub fn insert_padding(pieces: &[PieceInfo], sector_size: SectorSize) -> Result<Vec<PieceInfo>> {
+    let mut padded = Vec::new();
+    let mut written = 0;
+
+    for piece in pieces {
+        let PieceAlignment {
+            left_bytes,
+            right_bytes,
+        } = get_piece_alignment(UnpaddedBytesAmount(written), piece.size);
+
+        if left_bytes > UnpaddedBytesAmount(0) {
+            padded.extend(generate_padding_pieces(left_bytes)?);
+            written += u64::from(left_bytes);
+        }
+        padded.push(piece.clone());
+        written += u64::from(piece.size);
+
+        if right_bytes > UnpaddedBytesAmount(0) {
+            padded.extend(generate_padding_pieces(right_bytes)?);
+            written += u64::from(right_bytes);
+        }
+    }
+
+    let sector_unpadded: UnpaddedBytesAmount = sector_size.into();
+    let missing = u64::from(sector_unpadded) - written;
+    padded.extend(generate_padding_pieces(UnpaddedBytesAmount(missing))?);
+
+    Ok(padded)
+}
+
+lazy_static! {
+    pub(crate) static ref SINGLE_PADDING_COMMITMENT: [u8; 32] = {
+        let hash = blake2s_simd::blake2s(&[0u8; MINIMUM_PIECE_SIZE + 1]);
+        let mut res = [0u8; 32];
+        res.copy_from_slice(hash.as_ref());
+        res
+    };
+}
+
+fn generate_padding_pieces(size: UnpaddedBytesAmount) -> Result<Vec<PieceInfo>> {
+    ensure!(
+        u64::from(size) % MINIMUM_PIECE_SIZE as u64 == 0,
+        "Invalid size, must be a multiple of the node size"
+    );
+
+    let count = usize::from(size) / MINIMUM_PIECE_SIZE;
+
+    Ok(vec![
+        PieceInfo {
+            commitment: *SINGLE_PADDING_COMMITMENT,
+            size: UnpaddedBytesAmount(MINIMUM_PIECE_SIZE as u64)
+        };
+        count
+    ])
+}
```

Comment:
This works 'for now', but it can lead to a lot of wasteful hashing to keep recomputing padding pieces which are powers of two multiples (other than one) of the `MINIMUM_PIECE_SIZE`. I think we are going to need to perform those optimizations eventually, because this is a potentially expensive action that all verifiers need to perform — so we need to keep the cost to the minimum.

The simplest way to do this is probably just to precompute and cache the doubling so we have a table with all the powers-of-two. At most this will be 32 bytes * the height of our largest sector, and it would be fine to compute this at startup.

Here is an example of the issue, which hopefully highlights why we want to be able to look up arbitrary power-of-2-sized padding pieces in a table.

Supposedthe supplied pieces have sizes (in order) 11, 256.

We will need the following padding pieces between the two: 16 , 4, 1, 32, 64, 128

If we instead just pad with unit padding and allow the pieces to be generated during the piece-verification reduction, we do a lot of extra hashing. In this example, that's 4 + 2 + 0 + 5 + 6 + 7 = 24 hashes.